RavenDB-10677 ACID operations in cluster that includes documents

karmeli87 opened the pull request:
WIP

Sorry for the big commit.
If you prefer I can split it into client side and server side.

karmeli87 committed:
RavenDB-10677 ACID operations in cluster that includes documents

WIP

ayende commented on src/Raven.Client/Constants.cs:
```diff
@@ -40,6 +40,8 @@ private Headers()
 
             public const string ClientConfigurationEtag = "Client-Configuration-Etag";
 
+            public const string LastKnownClusterTransactionIndex = "LastKnownClusterTransactionIndex";
```
Use `Header-Case`. That is more readable and match other headers.
Better to use `Known-Raft-Index`

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -87,6 +89,20 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
 
+        public override Task<ResponseDisposeHandling> ProcessResponse(JsonOperationContext context, HttpCache cache, HttpResponseMessage response, string url)
+        {
+            if (response.Headers.TryGetValues("TransactionIndex", out var value))
```
It would be better to return this value explicitly in the response, rather than implicitly via headers

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -108,31 +124,36 @@ private void AppendOptions(StringBuilder sb)
+
+            if (_options.TryGetValue<ClusterBatchOptions>(out var clusterBatchOptions))
+            {
+                sb.Append("&clusterTransaction=true");
```
Shouldn't be something that we send via query parameters.
It should be a property of the batch itself.

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchOptions.cs:
```diff
@@ -1,20 +1,67 @@
-        public TimeSpan WaitForReplicasTimeout { get; set; }
-        public bool Majority { get; set; }
-        public bool ThrowOnTimeoutInWaitForReplicas { get; set; }
+        private Dictionary<Type, TValue> _dictionary;
```
Why did you make this change? 
The problem with this is that you have now added a few (expensive) dictionary look ups that to a common operation

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchOptions.cs:
```diff
@@ -1,20 +1,67 @@
+        public TimeSpan? RequestTimeout { get; set; }
+    }
+
+    public interface IBatchOptions
```
Why introduce an interface here?

ayende commented on src/Raven.Client/Documents/Commands/Batches/DeleteCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public string Id { get; }
+        public string Name => null;
+        public string ChangeVector => null;
```
This does not match the `ToJson` impl below

ayende commented on src/Raven.Client/Documents/Commands/Batches/DeleteCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,32 @@
+            {
+                [nameof(Id)] = Id,
+                [nameof(ChangeVector)] = Index.ToString(),
+                [nameof(Type)] = Type.ToString()
```
use `nameof` here

ayende commented on src/Raven.Client/Documents/Commands/Batches/DeleteCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,32 @@
+            return new DynamicJsonValue
+            {
+                [nameof(Id)] = Id,
+                [nameof(ChangeVector)] = Index.ToString(),
```
This is not a valid change vector

ayende commented on src/Raven.Client/Documents/Commands/Batches/PutCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public string Id { get; }
+        public string Name => null;
+        public string ChangeVector => null;
```
Doesn't match `ToJson` below.

ayende commented on src/Raven.Client/Documents/Commands/Batches/PutCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,35 @@
+            {
+                [nameof(Id)] = Id,
+                [nameof(Document)] = Document,
+                [nameof(ChangeVector)] = Index.ToString(),
```
Not a valid change vector

ayende commented on src/Raven.Client/Documents/Commands/Batches/PutCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,35 @@
+                [nameof(Id)] = Id,
+                [nameof(Document)] = Document,
+                [nameof(ChangeVector)] = Index.ToString(),
+                [nameof(Type)] = Type.ToString()
```
use `nameof`

ayende commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -148,7 +148,7 @@ public override IDocumentSession OpenSession(SessionOptions options)
             var databaseName = options.Database ?? Database;
             var requestExecutor = options.RequestExecutor ?? GetRequestExecutor(databaseName);
-            var session = new DocumentSession(databaseName, this, sessionId, requestExecutor);
+            var session = new DocumentSession(databaseName, this, sessionId, requestExecutor, options.TransactionMode);
```
Just pass the options object directly

ayende commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -342,7 +342,7 @@ private AsyncDocumentSession OpenAsyncSessionInternal(SessionOptions options)
             var databaseName = options.Database ?? Database;
             var requestExecutor = options.RequestExecutor ?? GetRequestExecutor(databaseName);
-            var session = new AsyncDocumentSession(databaseName, this, requestExecutor, sessionId);
+            var session = new AsyncDocumentSession(databaseName, this, requestExecutor, sessionId, options.TransactionMode);
```
Pass the options obj directly

ayende commented on src/Raven.Client/Documents/DocumentStoreBase.cs:
```diff
@@ -136,6 +136,34 @@ public string[] Urls
+            get
+
+            {
+                if (_lastTransactionIndex == 0)
```
Copy to a local variable first.

ayende commented on src/Raven.Client/Documents/IDocumentStore.cs:
```diff
@@ -168,6 +168,8 @@ public interface IDocumentStore : IDisposalNotification
 
         OperationExecutor Operations { get; }
 
+        long? LastTransactionIndex { get; set; }
```
That is misleading name. Transactions exists on local nodes as well.
And I'm not sure why we want to expose this in the public interface

ayende replied:
Needs to be removed from the public interface

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Operations/CompareExchange/GetCompareExchangeIndexOperation.cs:
```diff
@@ -0,0 +1,80 @@
+                pathBuilder.Append("/databases/")
+                    .Append(node.Database)
+                    .Append("/cmpxchg/indexes?")
+                    .Append("keys=")
```
Use multiple `key` arguments here.
That is a more natural way to go about it.

ayende commented on src/Raven.Client/Documents/Operations/CompareExchange/GetCompareExchangeIndexOperation.cs:
```diff
@@ -0,0 +1,80 @@
+{
+    public class GetCompareExchangeIndexResult
+    {
+        public long[] Indexes;
```
This is confusing, because you need to do matchups between the `kyes` and the `indexes`.
Return a dictionary with the matches.

ayende commented on src/Raven.Client/Documents/Session/AsyncDocumentSession.cs:
```diff
@@ -27,12 +27,13 @@ public partial class AsyncDocumentSession : InMemoryDocumentSessionOperations, I
             GenerateDocumentIdsOnStore = false;
             Attachments = new DocumentSessionAttachmentsAsync(this);
             Revisions = new DocumentSessionRevisionsAsync(this);
+            ClusterTransaction = new ClusterTransactionOperationAsync(this);
```
Do we really need to create a new obj every time? This is pretty rare.
Can do this lazily (also valid for the other ones here)

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+            {
+                if (Session.DeleteCompareExchange.ContainsKey(key) == false)
+                    Session.DeleteCompareExchange.Add(key, index.Value);
+                return;
```
You didn't remove it from `TrackedKeys` or check if the known index is the same.

ayende replied:
You also don't check `StoreCompareExchange`

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+                return;
+            }
+
+            var result = await Session.Operations.SendAsync(new GetCompareExchangeIndexOperation(key)).ConfigureAwait(false);
```
It is better to just ask the user to provide the `index` directly.
This way, they are aware that there is a remote call here, because they are the one making it.

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+        {
+            EnsureClusterSession();
+
+            Session.StoreCompareExchange[item.Key] = (item.Value, item.Index);
```
What about `DeleteCompareExchange`?

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+            if (Session.TransactionMode != TransactionMode.ClusterWide)
+            {
+                throw new InvalidOperationException($"This function is part of cluster transaction session, in order to use it you have to create the Session with '{nameof(TransactionMode.ClusterWide)}'.");
```
Add the full code that is required to create the session in this manner.

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+        private void EnsureClusterSession()
+        {
+            if (Marked)
```
what is marked? And why is this needed?

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+    public interface IClusterTransactionOperation
+    {
+        void DeleteCompareExchangeValue(string key, long? index = null);
```
* `index` is non optional
* need overload that take `CompareExchangeValue<T>`

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+    public interface IClusterTransactionOperationAsync
+    {
+        Task DeleteCompareExchangeValueAsync(string key, long? index = null);
```
Same

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+        public async Task DeleteCompareExchangeValueAsync(string key, long? index = null)
+        {
+            await DeleteCompareExchangeValueAsyncInternal(key, index).ConfigureAwait(false);
```
Why call `await` here, just return the task

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,180 @@
+
+        public async Task<CompareExchangeValue<T>> GetCompareExchangeValueAsync<T>(string key)
+        {
+            return await GetCompareExchangeValueAsyncInternal<T>(key).ConfigureAwait(false);
```
why call `await`?

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionAttachmentsBase.cs:
```diff
@@ -63,6 +63,11 @@ public void Store(string documentId, string name, Stream stream, string contentT
         {
+            if (Session.TransactionMode == TransactionMode.ClusterWide)
+            {
+                throw new NotSupportedException("This operation is not supported under cluster transaction");
```
Explain what is this operation that isn't supported.
"Storing Attachments are not supported as part of a cluster wide transaction..."

ayende replied:
And in general, throwing this should be done in the `PrepareForSaveChanges` or `ValidateClusterTx`

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -65,6 +68,9 @@ public abstract partial class InMemoryDocumentSessionOperations : IDisposable
         /// </summary>
         protected internal readonly HashSet<object> DeletedEntities = new HashSet<object>(ObjectReferenceEqualityComparer<object>.Default);
 
+        protected internal Dictionary<string, (object Item, long Index)> StoreCompareExchange;
```
Why do you have some state here and some in the `ClusterSessionBase `

ayende replied:
Why is the state spread around?

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -730,6 +741,50 @@ internal SaveChangesData PrepareForSaveChanges()
             return result;
         }
 
+        private void ValidateClusterTransaction()
```
Where is the actual validation here?

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -730,6 +741,50 @@ internal SaveChangesData PrepareForSaveChanges()
+                    {
+                        ["Object"] = entity.Value.Item
+                    };
+                    var blittable = EntityToBlittable.ConvertEntityToBlittable(tuple, Conventions, Context);
```
Why do you wrap this in an parent value?

ayende replied:
Why is this wrapped?

karmeli87 replied:
the value can be also a primitive, so I have to wrap it.

ayende commented on src/Raven.Client/Documents/Session/SessionOptions.cs:
```diff
@@ -2,9 +2,21 @@
+
+        /// <summary>
+        /// Once TransactionMode set 'ClusterWide' it  will perform the SaveChanges as a transactional cluster wide operation.
+        /// Any document load / store / delete will be part of this session's cluster transaction.
```
`load` will NOT be part of the cluster tx

ayende replied:
`load` isn't part of the tx

karmeli87 replied:
done

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -47,17 +48,12 @@ public DatabasesLandlord(ServerStore serverStore)
-        }
-
-        private void HandleClusterDatabaseChanged((string DatabaseName, long Index, string Type) t, ClusterDatabaseChangeType changeType)
+        private void HandleClusterDatabaseChanged((string DatabaseName, long Index, string Type, ClusterDatabaseChangeType ChangeType) t)
```
Why pass it as a value tuple here?

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -125,16 +121,28 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                            return;
+                        }
+
+                        task.ContinueWith(done =>
```
Error handling here it missing

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var database = task.Result;
```
Error handling here is missing

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                BatchHandler.ClusterTransactionMergedCommand mergedCommands;
+                try
+                {
+                    mergedCommands = new BatchHandler.ClusterTransactionMergedCommand(database, serverContext, index);
```
Why would just creating this throw?

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                catch (Exception e)
+                {
+                    rtnCtx.Dispose();
+                    database.ClusterTransactionWaiter.SetException(e);
```
Why do we need separate error handling here.

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+
+                        if (t.Exception != null)
+                        {
+                            database.ClusterTransactionWaiter.SetException(t.Exception.GetBaseException());
```
Why use the base exception?

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                                }
+                            }
+
+                            database.ClusterTransactionWaiter.SetResult(index, (mergedCommands.Reply, mergedCommands.Skipped, replicationTask, indexTask));
```
Please avoid using such complex value tuples.
Create real classes / structs. Otherwise the code get _complex_.

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,74 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                            return;
+                        }
+
+                        database.ClusterTransactionWaiter.SetException(new Exception());
```
What is this meant to do?

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,56 @@
+            old.SetResult(null);
+        }
+
+        public async Task<object> WaitForResult(long index)
```
`CancellationToken` is missing here.

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,56 @@
+            }
+        }
+
+        public async Task WaitForCompletion(long index)
```
`CancellationToken` is missing here.

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,56 @@
+    public class WaitForClusterTransactionToFinish
+    {
+        private TaskCompletionSource<object> _indexChanged = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly ConcurrentDictionary<long, object> _results = new ConcurrentDictionary<long, object>();
```
What is removing the values from this?

ayende replied:
Who cleans this up?

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -30,25 +32,59 @@ public class BatchHandler : DatabaseRequestHandler
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                    {
+                        context.Write(writer, new DynamicJsonValue
+                        {
```
Missing, the return raft index

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -30,25 +32,59 @@ public class BatchHandler : DatabaseRequestHandler
+                    {
+                        WaitForIndexesTimeout = waitForIndexesTimeout,
+                        WaitForIndexThrow = waitForIndexThrow,
+                        SpecifiedIndexesQueryString = specifiedIndexesQueryString.ToList(),
```
Do this only if you actually have a value.

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -85,6 +118,91 @@ public class BatchHandler : DatabaseRequestHandler
+            if ((bool)result.Result == false)
+            {
+                HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
+                throw new ConcurrencyException("Failed to perform compare exchange on {...}.");
```
Need to have a full error here.

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -85,6 +118,91 @@ public class BatchHandler : DatabaseRequestHandler
+                throw new ConcurrencyException("Failed to perform compare exchange on {...}.");
+            }
+
+            var reply = ((DynamicJsonArray Array, bool Skipped, Task ReplicationTask, Task IndexTask)) await Database.ClusterTransactionWaiter.WaitForResult(result.Index);
```
Don't use such complex value tuples

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -85,6 +118,91 @@ public class BatchHandler : DatabaseRequestHandler
+            var reply = ((DynamicJsonArray Array, bool Skipped, Task ReplicationTask, Task IndexTask)) await Database.ClusterTransactionWaiter.WaitForResult(result.Index);
+            if (reply.Skipped)
+            {
+                using (pool.AllocateOperationContext(out DocumentsOperationContext context))
```
Why are you returning this? What is the meaning of skipped in this context?

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -85,6 +118,91 @@ public class BatchHandler : DatabaseRequestHandler
+                await reply.ReplicationTask;
+            }
+
+            HttpContext.Response.Headers.Add("TransactionIndex", result.Index.ToString());
```
Return this explicitly in the result

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -261,27 +375,160 @@ private List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(HashSet<string>
+                // Check if we already have the documents of this cluster transaction.
+                if (currentRaftIndex >= _index)
+                {
+                    Skipped = true;
```
Okay, this isn't correct. But I'm not sure what should be done here.
The problem here is that you can't really return the current value, because it doesn't reflect the command, it might have already been updated again

ayende replied:
Still wrong

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -261,27 +375,160 @@ private List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(HashSet<string>
+
+                using (_serverContext.OpenReadTransaction())
+                {
+                    var first = Math.Max(ClusterTransactionCommand.ReadFirstIndex(_serverContext, Database.Name), currentRaftIndex + 1);
```
Why do you have this max here? The command id should _always_ be fixed.

ayende replied:
Still a question

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -261,27 +375,160 @@ private List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(HashSet<string>
+
+                        Options = command.Options;
+                        if (Options?.WaitForIndexesTimeout.HasValue == true)
+                        {
```
Why overwrite?

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -261,27 +375,160 @@ private List<Index> GetImpactedIndexesToWaitForToBecomeNonStale(HashSet<string>
+                                    AddDeleteResult(deleteResult, cmd.Id);
+                                    break;
+                                default:
+                                    break;
```
There should be an error here.

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -621,6 +621,12 @@ private static unsafe CommandType GetCommandType(JsonParserState state, JsonOper
                     return CommandType.AttachmentDELETE;
 
+                case 18:
+                    return CommandType.CompareExchangePUT;
```
You are doing _length_ comparison. You need to actually check the value.

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -621,6 +621,12 @@ private static unsafe CommandType GetCommandType(JsonParserState state, JsonOper
+                    return CommandType.CompareExchangePUT;
+
+                case 20:
+                    return CommandType.CompareExchangeDELETE;
```
Same.

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -953,6 +956,7 @@ public void AcceptNewConnection(Stream stream, Action disconnect, EndPoint remot
                         }
                     }
                     _clusterId = initialMessage.TopologyId;
+                    SetClusterBase(_clusterId);
```
There are a bunch of more places where this is used, for example, when the value is changed on the fly using admin js console

ayende replied:
Still need to add this in more places

ayende commented on src/Raven.Server/Routing/RequestRouter.cs:
```diff
@@ -102,7 +103,13 @@ public RouteInformation GetRoute(string method, string path, out RouteMatch matc
+                {
+                    if (reqCtx.HttpContext.Response.Headers.TryGetValue(Constants.Headers.LastKnownClusterTransactionIndex, out var value) && long.TryParse(value, out var index))
+                    {
+                        await reqCtx.Database.ClusterTransactionWaiter.WaitForCompletion(index);
```
Before going to the `await`, do the check directly there. It can save some work

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -290,6 +306,24 @@ protected override void Apply(TransactionOperationContext context, BlittableJson
             }
         }
 
+        private void ExecuteClusterTransaction(TransactionOperationContext context, BlittableJsonReaderObject cmd, long index, out bool success)
```
Avoid using out variable when you can use return code.
And using `bool` here doesn't help much. We need to return an explicit explanation of what failed if we weren't able to apply things.

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -551,6 +585,12 @@ private void DeleteDatabaseRecord(TransactionOperationContext context, long inde
+            var transactionsCommands = context.Transaction.InnerTransaction.OpenTable(TransactionCommandsSchema, TransactionCommands);
+            using (Slice.From(context.Allocator, databaseName, out var loweredKey))
+            {
+                transactionsCommands.DeleteByKey(loweredKey);
```
This is wrong. You are trying to delete where key = 'loweredKey`
But you can have more than one such command

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1056,7 +1078,24 @@ private void OnTransactionDispose(TransactionOperationContext context, long inde
         }
-        
+
+        public IEnumerable<long> GetCompareExchangeIndexes(TransactionOperationContext context, string[] keys)
```
Better to return a dictionary here

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+        public List<CompareExchangeCommand> ClusterCommands = new List<CompareExchangeCommand>();
+        public BlittableJsonReaderObject SerializedDatabaseCommands;
+
+        [JsonDeserializationIgnore]
```
Where are we using json serialization on this?

ayende replied:
why JSON serialization?

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+
+            foreach (var command in commandParsedCommands)
+            {
+                if (command.Type == CommandType.PUT || command.Type == CommandType.DELETE)
```
switch case would be easier here

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+                    var put = new AddOrUpdateCompareExchangeCommand(clusterCommand.Key, clusterCommand.Item, clusterCommand.Index, context);
+                    if (put.Validate(context, items, index) == false)
+                    {
+                        return false;
```
Need to create explanation why

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+        {
+            var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+            var commands = SerializedDatabaseCommands.Clone(context);
+            using (Slice.From(context.Allocator, Database + "/", out Slice databaseSlice))
```
Better to use record separator (ASCII code 30) here.

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+            databaseBytes.CopyTo(prefix, 0);
+            indexBytes.CopyTo(prefix, databaseBytes.Length);
+
+            using (Slice.From(context.Allocator, prefix, out Slice prefixSlice))
```
This uses a lot of allocations that aren't required. Can do that directly on a buffer.

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+
+            using (Slice.From(context.Allocator, prefix, out Slice prefixSlice))
+            {
+                var commandsBluk = items.SeekForwardFrom(ClusterStateMachine.TransactionCommandsSchema.Indexes[ClusterStateMachine.CommandByDatabaseAndIndex],
```
`Bluk` ?
`Single()` - That will error a lot

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,260 @@
+        public override DynamicJsonValue ToJson(JsonOperationContext context)
+        {
+            var djv = base.ToJson(context);
+            djv[nameof(ClusterCommands)] = new DynamicJsonArray(ClusterCommands.Select(x => x.ToJson()));
```
I'm confused about what you are doing here.

ayende commented on src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:
```diff
@@ -31,6 +31,20 @@ protected CompareExchangeCommandBase(string key, long index, JsonOperationContex
+                    var itemIndex = *(long*)reader.Read((int)ClusterStateMachine.UniqueItems.Index, out var _);
+                    return Index == itemIndex;
+                }
+            }
```
You are missing handling for missing value.
Better to use the `Slice` ability to do lower casing, instead.

ayende commented on src/Raven.Server/Utils/ChangeVectorUtils.cs:
```diff
@@ -241,14 +242,25 @@ public static string NewChangeVector(string nodeTag, long etag, string dbIdInBas
                 .ToString();
         }
 
-        public static long GetEtagByNodeTag(string changeVector, string nodeTag)
```
Did no one used `ByNodeTag` ?

karmeli87 replied:
This is wrong to get the Etag by the node, since it is not unique and could come from different database or even cluster.

ayende commented on src/Raven.Server/Utils/ChangeVectorUtils.cs:
```diff
@@ -241,14 +242,25 @@ public static string NewChangeVector(string nodeTag, long etag, string dbIdInBas
-            var etag = changeVector.IndexOf('-', index);
-            return long.Parse(changeVector.Substring(index + offset, etag - index - offset));
+
+            var bytes = Encoding.UTF8.GetBytes(changeVector);
```
why are you doing this scanning on the byte representation?

ayende replied:
Why not use `LastIndexOf` ?

ayende commented on src/Raven.Server/Utils/TypeConverter.cs:
```diff
@@ -64,6 +64,20 @@ public static bool IsSupportedType(object value)
 
             List<bool> supporeted = new List<bool>();
```
Let's kill this in favor of a simple early break

ayende commented on src/Raven.Server/Web/System/CompareExchangeHandler.cs:
```diff
@@ -34,6 +34,29 @@ public Task GetCompareExchangeValues()
+        [RavenAction("/databases/*/cmpxchg/indexes", "GET", AuthorizationStatus.ValidUser)]
+        public Task GetCompareExchangeIndexes()
+        {
+            var keys = GetStringValuesQueryString("keys");
```
Use the query string ability to send mulitple values to avoid doing string splits

ayende commented on src/Sparrow/Json/JsonOperationContext.cs:
```diff
@@ -553,6 +553,25 @@ public BlittableJsonReaderObject ReadForMemory(Stream stream, string documentId)
             return ReadObjectInternal(builder, documentId, mode, modifier);
         }
 
+        public BlittableJsonReaderArray ReadArray(DynamicJsonArray builder, string documentId,
```
Why do you need this? We almost never get or send arrays directly, to have better versionning support

karmeli87 committed:
RavenDB-10677 ACID operations in cluster that includes documents

Still very much WIP. fix most PR comments

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -62,6 +63,12 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
+                        if (_options?.TransactionMode == TransactionMode.ClusterWide)
+                        {
+                            writer.WriteComma();
+                            writer.WritePropertyName("ClusterTx");
```
`'TransactionMode': 'ClusterWide'` is better.

ayende commented on src/Raven.Client/Documents/Operations/CompareExchange/GetCompareExchangeIndexOperation.cs:
```diff
@@ -68,11 +70,23 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
+                {
+                    if (response.TryGetMember(propertyName, out object val))
+                    {
+                        Result.Add(propertyName, (long)val);
```
This will throw NRE here.

ayende commented on src/Raven.Client/Documents/Session/AsyncDocumentSession.cs:
```diff
@@ -99,11 +99,14 @@ protected override Task<string> GenerateIdAsync(object entity)
 
-        public IAttachmentsSessionOperationsAsync Attachments { get; }
+        public IAttachmentsSessionOperationsAsync Attachments => _attachments.Value;
+        private readonly Lazy<IAttachmentsSessionOperationsAsync> _attachments;
```
Avoid the `Lazy` here, we don't need thread safety.
We can just use:  

```
public IAttachmentsSessionOperationsAsync Attachments => _attachments ?? (_attachments = new ...);`
```

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -8,105 +9,87 @@ namespace Raven.Client.Documents.Session
         }
 
-        protected void CreateCompareExchangeValueInternal<T>(string key, T item)
+        public void CreateCompareExchangeValue<T>(string key, T item)
```
Missing check for `DeletedCompareExchange`

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -8,105 +9,87 @@ namespace Raven.Client.Documents.Session
-                    Session.DeleteCompareExchange.Add(key, index.Value);
-                return;
-            }
+            if (Session.DeleteCompareExchange.ContainsKey(item.Key) == false)
```
Missing check for `StoreCompareExchange`.

Where are you checking that we aren't trying to send conflicting commands?

ayende commented on src/Raven.Client/Documents/Session/DocumentSession.cs:
```diff
@@ -46,27 +46,30 @@ public partial class DocumentSession : InMemoryDocumentSessionOperations, IDocum
-        public IAttachmentsSessionOperations Attachments { get; }
-        
+        public IAttachmentsSessionOperations Attachments => _attachments.Value;
+        private readonly Lazy<IAttachmentsSessionOperations> _attachments;
```
Same comment about not using `Lazy`

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -74,25 +76,25 @@ public void SetResult(long index, object result)
             while (true)
             {
                 var tcs = _indexChanged;
                 if (_lastIndex >= index)
```
Also check cancellation here

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -74,25 +76,25 @@ public void SetResult(long index, object result)
+        public async Task<object> WaitForResult(long index, CancellationToken token = default)
         {
             while (true)
             {
```
Also check cancellation here

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -34,22 +35,21 @@ public class BatchHandler : DatabaseRequestHandler
                     contentType.StartsWith("application/json", StringComparison.OrdinalIgnoreCase))
                 {
-                    command.ParsedCommands = await BatchRequestParser.BuildCommandsAsync(context, RequestBodyStream(), Database, ServerStore);
+                    isClusterTransaction = await BatchRequestParser.BuildCommandsAsync(context, command, RequestBodyStream(), Database, ServerStore);
```
Bad API.
Create a field on the `command` object instead

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -34,22 +35,21 @@ public class BatchHandler : DatabaseRequestHandler
                     contentType.StartsWith("multipart/form-data", StringComparison.OrdinalIgnoreCase))
                 {
-                    await ParseMultipart(context, command);
+                    isClusterTransaction = await ParseMultipart(context, command);
```
Bad API.

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -199,21 +208,21 @@ public class BatchHandler : DatabaseRequestHandler
         }
 
-        private async Task ParseMultipart(DocumentsOperationContext context, MergedBatchCommand command)
+        private async Task<bool> ParseMultipart(DocumentsOperationContext context, MergedBatchCommand command)
```
This is a bad API, what does `bool` mean, success flag, usually

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -75,7 +75,8 @@ private static void ReturnBuffer(CommandData[] cmds)
 
-        public static async Task<ArraySegment<CommandData>> BuildCommandsAsync(JsonOperationContext ctx, Stream stream, DocumentDatabase database, ServerStore serverStore)
+        public static async Task<bool> BuildCommandsAsync(JsonOperationContext ctx, BatchHandler.MergedBatchCommand command, Stream stream,
+            DocumentDatabase database, ServerStore serverStore)
```
Same issue with API

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -161,8 +162,33 @@ private static void ReturnBuffer(CommandData[] cmds)
+        private static unsafe bool ReadClusterTransactionProperty(JsonParserState state)
+        {
+            return state.CurrentTokenType == JsonParserToken.String &&
+                   GetLongFromStringBuffer(state) == 6085037597358648387 && // ClusterT
```
You are not checking the size of the string here, can cause invalid memory access

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -622,9 +648,22 @@ private static unsafe CommandType GetCommandType(JsonParserState state, JsonOper
+                    }
                     return CommandType.CompareExchangePUT;
 
                 case 20:
```
The size is `21`, not `20`

ayende commented on src/Raven.Server/Routing/RequestRouter.cs:
```diff
@@ -106,7 +106,7 @@ public RouteInformation GetRoute(string method, string path, out RouteMatch matc
                     if (reqCtx.HttpContext.Response.Headers.TryGetValue(Constants.Headers.LastKnownClusterTransactionIndex, out var value) && long.TryParse(value, out var index))
                     {
-                        await reqCtx.Database.ClusterTransactionWaiter.WaitForCompletion(index);
+                        await reqCtx.Database.ClusterTransactionWaiter.WaitForCompletion(index, reqCtx.HttpContext.RequestAborted);
```
Try early exit here first

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -103,74 +103,88 @@ public ClusterTransactionCommand(string database, ArraySegment<BatchRequestParse
+                        {
+                            Key = command.Id,
+                            Type = command.Type,
+                            Index = long.Parse(command.ChangeVector),
```
Wrong, the change vector should _be_ a change vector

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -201,25 +215,42 @@ public static unsafe long ReadFirstIndex(TransactionOperationContext serverConte
+        private static byte[] GetPrefix(string database, long? index = null)
         {
-            var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+            var databaseBytes = Encoding.UTF8.GetBytes(database);
```
There is no need to do this. Use a single buffer here.

ayende commented on src/Raven.Server/Web/System/CompareExchangeHandler.cs:
```diff
@@ -37,21 +37,19 @@ public Task GetCompareExchangeValues()
+                foreach (var tuple in ServerStore.Cluster.GetCompareExchangeIndexes(context, keys))
+                {
+                    writer.WritePropertyName(tuple.Key);
+                    writer.WriteInteger(tuple.Index);
```
Missing `WriteComma`

ayende commented:
Also, need to fix the confclits

karmeli87 committed:
RavenDB-10677 ACID operations in cluster that includes documents

finishing implementation.

ayende commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -60,6 +63,12 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
+                        {
+                            writer.WriteComma();
+                            writer.WritePropertyName(nameof(TransactionMode));
+                            writer.WriteString(TransactionMode.ClusterWide.ToString());
```
`nameof(ClusterWide)`

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Commands/Batches/DeleteCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,32 @@
+            {
+                [nameof(Id)] = Id,
+                [nameof(Index)] = Index,
+                [nameof(Type)] = Type.ToString(),
```
static, use `nameof`

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Commands/Batches/PutCompareExchangeCommandData.cs:
```diff
@@ -0,0 +1,35 @@
+                [nameof(Id)] = Id,
+                [nameof(Document)] = Document,
+                [nameof(Index)] = Index,
+                [nameof(Type)] = Type.ToString()
```
static, use `nameof`

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Operations/CompareExchange/GetCompareExchangeIndexOperation.cs:
```diff
@@ -0,0 +1,94 @@
+
+                foreach (var propertyName in response.GetPropertyNames())
+                {
+                    if (response.TryGetMember(propertyName, out object val))
```
`out long val`

karmeli87 replied:
done

ayende commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -0,0 +1,165 @@
+        {
+            try
+            {
+                return Session.Operations.SendAsync(new GetCompareExchangeValueOperation<T>(key));
```
This is wrong. You want to add additional exception handling, but you forgot to `await`

karmeli87 replied:
OK, I will simply to return the Task

ayende commented on src/Raven.Client/Documents/Session/Operations/BatchOperation.cs:
```diff
@@ -27,15 +27,19 @@ public BatchCommand CreateRequest()
-        public void SetResult(BlittableArrayResult result)
+        public void SetResult(BatchCommandResult result)
         {
+            _session.DocumentStore.LastTransactionIndex = result.TransactionIndex;
```
This should be a method, not a property

karmeli87 replied:
done

ayende commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -646,6 +646,11 @@ protected virtual bool TryLoadFromCache(JsonOperationContext context)
                 if (_disableClientConfigurationUpdates == false)
                     request.Headers.TryAddWithoutValidation(Constants.Headers.ClientConfigurationEtag, $"\"{ClientConfigurationEtag.ToInvariantString()}\"");
 
+                if (sessionInfo?.LastClusterTransaction.HasValue == true)
```
`!= null` reads better here.

karmeli87 replied:
done

ayende commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -65,6 +65,8 @@ public DatabaseRecord(string databaseName)
 
         public ClientConfiguration Client;
 
+        public long TrunctedClusterTransactionIndex;
```
Truncated

karmeli87 replied:
done

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -146,6 +154,68 @@ private void HandleClusterDatabaseChanged((string DatabaseName, long Index, stri
+                return;
+            }
+
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
```
What happens if this throws? Why not wrap the whole thing in tx?

karmeli87 replied:
It is a problem if this throws... Can we do something about it?
We can't even read the task id in order to notify an exception.

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,73 @@
+        public void SetException(Exception e)
+        {
+            var old = Interlocked.Exchange(ref _indexChanged, new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously));
+            old.SetException(e);
```
This is racy with respect to waiting clients

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,73 @@
+            ClusterStateMachine.InterlockedExchangeMax(ref _lastIndex, index);
+
+            var old = Interlocked.Exchange(ref _indexChanged, new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously));
+            old.SetResult(null);
```
This is racy

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -52,6 +53,73 @@
+            old.SetException(e);
+        }
+
+        public void CleanUp(long recordTrunctedClusterIndex)
```
Who calls this?

ayende replied:
And why are we waiting until the _cluster_ decide to do something about it?

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -48,8 +49,33 @@ public class BatchHandler : DatabaseRequestHandler
 
                 var waitForIndexesTimeout = GetTimeSpanQueryString("waitForIndexesTimeout", required: false);
+                var waitForIndexThrow = GetBoolValueQueryString("waitForIndexThrow", required: false) ?? true;
+                var specifiedIndexesQueryString = HttpContext.Request.Query["waitForSpecificIndex"];
```
Add comment on why we aren't waiting for replication here

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -64,28 +90,115 @@ public class BatchHandler : DatabaseRequestHandler
+
+            var reply = (ClusterTransactionCompletionResult)await Database.ClusterTransactionWaiter.WaitForResult(result.Index, HttpContext.RequestAborted);
+            if (reply.Skipped)
+            {
```
All of this is wrong.

ayende commented on src/Raven.Server/Documents/Handlers/BatchHandler.cs:
```diff
@@ -304,17 +570,20 @@ private bool CanAvoidThrowingToMerger(ConcurrencyException e, int commandOffset)
-                var counterBatch = new CounterBatch();
-                Reply = new DynamicJsonArray();
-                for (int i = ParsedCommands.Offset; i < ParsedCommands.Count; i++)
+var counterBatch = new CounterBatch();                for (int i = ParsedCommands.Offset; i < ParsedCommands.Count; i++)
```
strange formatting

ayende commented on src/Raven.Server/Documents/Handlers/BatchRequestParser.cs:
```diff
@@ -169,8 +170,36 @@ private static void ReturnBuffer(CommandData[] cmds)
+                while (parser.Read() == false)
+                    await RefillParserBuffer(stream, buffer, parser);
+
+                return GetStringPropertyValue(state) == TransactionMode.ClusterWide.ToString();
```
`nameof`

ayende commented on src/Raven.Server/Routing/RequestRouter.cs:
```diff
@@ -102,7 +103,14 @@ public RouteInformation GetRoute(string method, string path, out RouteMatch matc
+                    if (reqCtx.HttpContext.Response.Headers.TryGetValue(Constants.Headers.LastKnownClusterTransactionIndex, out var value) && long.TryParse(value, out var index))
+                    {
+                        reqCtx.HttpContext.RequestAborted.ThrowIfCancellationRequested();
+                        await reqCtx.Database.ClusterTransactionWaiter.WaitForCompletion(index, reqCtx.HttpContext.RequestAborted);
```
Do this eagerly, not lazily

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -535,6 +624,12 @@ private void DeleteDatabaseRecord(TransactionOperationContext context, long inde
+            var transactionsCommands = context.Transaction.InnerTransaction.OpenTable(TransactionCommandsSchema, TransactionCommands);
+            using (Slice.From(context.Allocator, databaseName, out var loweredKey))
+            {
+                transactionsCommands.DeleteForwardFrom(TransactionCommandsSchema.Indexes[CommandByDatabaseAndIndex], loweredKey, true, long.MaxValue);
```
This will delete commands for other databases

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,314 @@
+
+            using (Slice.From(serverContext.Allocator, GetPrefix(database), out var prefixSlice))
+            {
+                var reader = items.SeekOneForwardFrom(ClusterStateMachine.TransactionCommandsSchema.Indexes[ClusterStateMachine.CommandByDatabaseAndIndex], prefixSlice);
```
Will get it from wrong database

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -0,0 +1,314 @@
+            var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+            using (Slice.From(context.Allocator, GetPrefix(database, index), out Slice prefixSlice))
+            {
+                var commandsBulk = items.SeekForwardFrom(ClusterStateMachine.TransactionCommandsSchema.Indexes[ClusterStateMachine.CommandByDatabaseAndIndex],
```
WIll use wrong db

ayende commented on src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:
```diff
@@ -31,6 +31,21 @@ protected CompareExchangeCommandBase(string key, long index, JsonOperationContex
+            currentIndex = -1;
+            using (Slice.From(context.Allocator, dbKey, out Slice keySlice))
+            {
+                if (items.ReadByKey(keySlice, out var reader))
```
What if the value doesn't not exists ?

karmeli87 replied:
will return `true` and currentIndex `-1`

ayende commented on src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs:
```diff
@@ -165,6 +166,16 @@ class ClusterObserver : IDisposable
 
                             updateCommands.Add((cmd, updateReason));
                         }
+
```
`cleanupIndex` is a bad name, I thought you refer to indexes

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -225,6 +225,10 @@ public long LastRaftCommitIndex
                                         if (notification == null)
                                             break;
+
+                                        if (notification.TryGet("Id",out string id) && id != nameof(ClusterTopologyChanged))
```
What is this about?

karmeli87 replied:
It seems that even if the notification is of a different type (but same base type) it still will serialize it in a bad manner.

ayende commented on src/Raven.Server/Web/System/CompareExchangeHandler.cs:
```diff
@@ -34,6 +34,32 @@ public Task GetCompareExchangeValues()
         }
 
+        [RavenAction("/databases/*/cmpxchg/indexes", "GET", AuthorizationStatus.ValidUser)]
+        public Task GetCompareExchangeIndexes()
```
what is this for?

karmeli87 replied:
If I want to update several compare exchange values.. So I need to know only their indexes.

karmeli87 committed:
RavenDB-10677 ACID operations in cluster that includes documents

WIP, Fixing PR

ayende committed:
Merge branch 'RavenDB-10677' of https://github.com/karmeli87/ravendb into karmeli87-RavenDB-10677

# Conflicts:
#	src/Raven.Server/Documents/Handlers/BatchHandler.cs
#	test/Tryouts/Program.cs

ayende reviewed (commented):
protected override ClusterTransactionSessionBase GetClusterSession()
SetTransactionMode(ClusterWide);


Debug.Assert(cmd.IdPrefixed == false);
 private static readonly byte[] CommandsCountKey = Encoding.UTF8.GetBytes("CommandsCount");

ayende committed:
RavenDB-10677 - Core review changes

aviviadi committed:
RavenDB-11099: Clear cache and reset context on heavily variance between documents' properties (#6158)

* RavenDB-11099: Clear cache and reset context on heavily variance between documents' properties

ayende committed:
RavenDB-11089 - Seperating entities serialization (user controls that) and command serialization (must use default serialization)

ayende committed:
RavenDB-11089 - Add the new command serialization

gregolsky committed:
RavenDB-11033 require PS version 4.0

gregolsky committed:
RavenDB-11103 Unit testing - Unable to start server

gregolsky committed:
RavenDB-9321 Confusing message when setting up Docker

aviviadi committed:
RavenDB-11099: StreamOperation Async - Renew CachedProperties

gregolsky committed:
RavenDB-11033 rename property

gregolsky committed:
RavenDB-11033 use temp dir instead of base dir

karmeli87 committed:
RavenDB-10677

Fixing PR comments

ppekrol committed:
RavenDB-8766 unify logging sources

ppekrol committed:
RavenDB-10483 fixed concurrent requests count calculation

ayende committed:
RavenDB-11089 - Fixing query parameters need to use the user's conventions

Commit:
RavenDB-9196 Added escaping to BlittablePath now if you want to support paths like 'i like indexers[]' you should use the path 'i like indexers\[\]'

ppekrol committed:
RavenDB-11073 session.Advanced.Attachments.GetNames should throw is entity is not bound to session

ayende committed:
RDBC-163 Adding passing test

ayende committed:
RavenDB-11111 Query cache ignores null values

ppekrol committed:
RavenDB-11073 fixing test

ayende commented on src/Raven.Client/Documents/Session/IAdvancedDocumentSessionOperations.cs:
```diff
@@ -170,7 +170,13 @@ public interface IAdvancedDocumentSessionOperations
+        /// Overwrite the existing transaction mode
+        /// </summary>
+        /// <param name="mode"></param>
+        void SetTransactionMode(TransactionMode mode);
```
This is validated on `SaveChanges`, IIRC. So if we set it to normal but we have added cmpxchng, we'll throw.

karmeli87 replied:
Will add a test to verify the behavior.

ayende commented on src/Raven.Server/Documents/ConflictsStorage.cs:
```diff
@@ -815,37 +816,62 @@ private static int FindIndexOfLargestEtagAndMergeChangeVectors(IReadOnlyList<Doc
             else
                 return ConflictStatus.Update; //document with 'id' doesn't exist locally, so just do PUT
 
+            if (hasLocalClusterTx || hasRemoteClusterTx)
```
This might be too early

karmeli87 replied:
This is only to mark that we need to remove the ClusterTx flag when we will put the document.

ayende replied:
Needs to be moved _after_ you know that this isn't already okay

ayende commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -183,7 +183,7 @@ private void NotifyPendingClusterTransaction(string databaseName, Task<DocumentD
                     var firstCommandIndex = ClusterTransactionCommand.ReadFirstIndex(serverContext, database.Name);
                     var first = Math.Max(firstCommandIndex, currentRaftIndex + 1);
-                    foreach (var command in ClusterTransactionCommand.ReadCommandsBatch(serverContext, database.Name, first))
+                    foreach (var command in ClusterTransactionCommand.ReadCommandsBatch(serverContext, database.Name.ToLowerInvariant(), first))
```
Shouldn't the `ToLower` be inside the `ReadCommandsBatch`?

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1100,7 +1101,7 @@ private static DocumentTombstone TableValueToTombstone(JsonOperationContext cont
                     modifiedTicks,
                     changeVector,
-                    local.Tombstone.Flags | documentFlags).Etag;
+                    localFlags | documentFlags).Etag;
```
why not use `flags` here?

karmeli87 replied:
done

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1122,7 +1121,7 @@ public override int Execute(DocumentsOperationContext context)
 #endif
                                                 database.DocumentsStorage.Put(context, item.Id, null, document, item.LastModifiedTicks, rcvdChangeVector,
-                                                    item.Flags, NonPersistentDocumentFlags.FromReplication);
+                                                    item.Flags.Strip(DocumentFlags.FromClusterTransaction), NonPersistentDocumentFlags.FromReplication);
```
Why is this stripped here?

karmeli87 replied:
You are right.

need to strip it only if `removeClusterTx == true`.

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1133,20 +1132,64 @@ public override int Execute(DocumentsOperationContext context)
                                                         new CollectionName(item.Collection),
-                                                        NonPersistentDocumentFlags.FromReplication);
+                                                        NonPersistentDocumentFlags.FromReplication, 
+                                                        item.Flags.Strip(DocumentFlags.FromClusterTransaction));
```
Why is this stripped here?

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1133,20 +1132,64 @@ public override int Execute(DocumentsOperationContext context)
                                             break;
                                         case ConflictStatus.AlreadyMerged:
-                                            //nothing to do
+                                            if (removeClusterTx)
```
What is the scenario where this actually happens, and what are the effects?

karmeli87 replied:
We can get here from the conflict status. When the local groupId is higher than the remote one.
The scenario that I can think of: an external replication of cluster transaction to database that also performed cluster transaction.

So at first will resolve it to the local document. (and stripping the FromCluster flag)

ayende commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -232,15 +232,15 @@ private void UpdateScriptResolvers()
                 if (incoming.Doc != null)
                 {
-                    _database.DocumentsStorage.RevisionsStorage.Put(context, incoming.Id, incoming.Doc, incoming.Flags | DocumentFlags.Conflicted | DocumentFlags.HasRevisions,
+                    _database.DocumentsStorage.RevisionsStorage.Put(context, incoming.Id, incoming.Doc, incoming.Flags.Strip(DocumentFlags.FromClusterTransaction) | DocumentFlags.Conflicted | DocumentFlags.HasRevisions,
```
Why are we removing important info (that the revision is from cluster tx) ?

karmeli87 replied:
You are right. Was too eager about removing it..

ayende commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -232,15 +232,15 @@ private void UpdateScriptResolvers()
                     {
                         _database.DocumentsStorage.RevisionsStorage.Delete(context, incoming.Id, key, new CollectionName(incoming.Collection), newChangeVector,
-                            incoming.LastModified.Ticks, NonPersistentDocumentFlags.None, incoming.Flags | DocumentFlags.Conflicted | DocumentFlags.HasRevisions);
+                            incoming.LastModified.Ticks, NonPersistentDocumentFlags.None, incoming.Flags.Strip(DocumentFlags.FromClusterTransaction) | DocumentFlags.Conflicted | DocumentFlags.HasRevisions);
```
Why are we removing important info here?

ayende commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -251,15 +251,15 @@ private void UpdateScriptResolvers()
                 {
                     _database.DocumentsStorage.RevisionsStorage.Put(context, existing.Document.Id, existing.Document.Data,
-                        existing.Document.Flags | DocumentFlags.Conflicted | DocumentFlags.HasRevisions,
+                        existing.Document.Flags.Strip(DocumentFlags.FromClusterTransaction) | DocumentFlags.Conflicted | DocumentFlags.HasRevisions,
```
Why is this removed?

ayende commented on src/Raven.Server/Documents/Replication/ResolveConflictOnReplicationConfigurationChange.cs:
```diff
@@ -251,15 +251,15 @@ private void UpdateScriptResolvers()
                     {
                         _database.DocumentsStorage.RevisionsStorage.Delete(context, existing.Tombstone.LowerId, key, new CollectionName(existing.Tombstone.Collection), existing.Tombstone.ChangeVector,
-                            existing.Tombstone.LastModified.Ticks, NonPersistentDocumentFlags.None, existing.Tombstone.Flags | DocumentFlags.Conflicted | DocumentFlags.HasRevisions);
+                            existing.Tombstone.LastModified.Ticks, NonPersistentDocumentFlags.None, existing.Tombstone.Flags.Strip(DocumentFlags.FromClusterTransaction) | DocumentFlags.Conflicted | DocumentFlags.HasRevisions);
```
Why is this removed?

ayende commented on src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:
```diff
@@ -27,18 +26,21 @@ public class ClusterTransactionDataCommand
             {
+                if (command.IdPrefixed)
+                {
+                    throw new NotSupportedException("Deleting by prefix is not supported.");
```
In what context? "deleting by prefix is not supported on cluster transaction... "

karmeli87 replied:
done.

ayende commented on src/Raven.Server/Utils/ChangeVectorUtils.cs:
```diff
@@ -81,92 +80,8 @@ public static ConflictStatus GetConflictStatus(string remoteAsString, string loc
             return remoteHasLargerEntries ? ConflictStatus.Update : ConflictStatus.AlreadyMerged;
         }
 
-        public static ConflictStatus GetConflictStatus(string remoteAsString, string localAsString, string priority, bool clusterAutoResolve = true)
```
This is a whole lot of code that is going away +1

ppekrol committed:
RavenDB-10581 when we are using / in the ID, then we need to check if the generated ID does not overlap already existing one

ppekrol committed:
RavenDB-10581 introduced OverlappingGeneratedNewIdException

karmeli87 committed:
RavenDB-10677

Fixing PR comments

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1115,13 +1115,20 @@ public override int Execute(DocumentsOperationContext context)
+                                            var flags = item.Flags;
+                                            if (removeClusterTx)
+                                            {
+                                                flags.Strip(DocumentFlags.FromClusterTransaction);
```
flags =

arekpalinski committed:
RavenDB-10982 Fixing race condition in the test (errors are stored in separate tx)

ayende committed:
RavenDB-10581
 - Moved the state of retries to the calling commands.
 - Command can now ask to be retried (once)
 - Will work even on two commands where the first failed and the second also have issues.

ppekrol committed:
RavenDB-10978 Create OfflineMigrationOperation

ayende committed:
RavenDB-10581 - Reverting other changes, avoid infinite loop on error, removed unused exception type

ayende committed:
RavenDB-10581 Avoid repeating myself and ensuring that we use the same format as expected of us

Danielle9897 committed:
RavenDB-11110 Spatial Index with 'BoundingBox'

ppekrol committed:
RavenDB-10581 fixing compilation

ppekrol committed:
RavenDB-10978 fixing Studio compilation

karmeli87 committed:
RavenDB-10677

Fixing PR comments

ppekrol committed:
RavenDB-7070 minor rename

ayende commented on src/Raven.Server/Rachis/Leader.cs:
```diff
@@ -788,6 +788,11 @@ public enum TopologyModification
 
         public bool TryModifyTopology(string nodeTag, string nodeUrl, TopologyModification modification, out Task task, bool validateNotInTopology = false, Action<TransactionOperationContext> beforeCommit = null)
         {
+            if (nodeTag != null && nodeTag.Equals("RAFT"))
```
It will require 298,941 nodes in the cluster to hit :-)

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1176,7 +1173,7 @@ public override int Execute(DocumentsOperationContext context)
                                             break;
                                         case ConflictStatus.AlreadyMerged:
-                                            if (removeClusterTx)
+                                            if (shouldStripClusterTxFlag)
```
It will be clearer if you'll move it higher.

ayende committed:
Merge branch 'RavenDB-10677' of https://github.com/karmeli87/ravendb into karmeli87-RavenDB-10677

# Conflicts:
#	src/Raven.Server/Routing/RequestRouter.cs

ayende committed:
RavenDB-10677 Fixing complication

ayende committed:
RavenDB-10677 Refactoring to avoid code duplication and to reduce size of internal state

ayende commented:
@karmeli87 Please review 6b63277, the logic should stay the same, but I reduced the amount of state you have to keep in your head to follow the code

ayende committed:
RavenDB-10677 - Fixing NRE when executing update command

ayende committed:
RavenDB-10677 Fixing missing db when applying cmpxchng commands

karmeli87 committed:
RavenDB-10677

- Fixing compare exchange export/import
- Fixing compare exchange queries.

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1141,7 +1140,7 @@ public IEnumerable<(string Key, long Index)> GetCompareExchangeIndexes(Transacti
         {
             var ptr = reader.Read((int)UniqueItems.Key, out var size);
-            return Encodings.Utf8.GetString(ptr, size).Substring(dbPrefix.Length + 1);
+            return Encodings.Utf8.GetString(ptr, size).Substring(dbPrefix.Length + 4);
```
What is the meaning here?

karmeli87 replied:
We add the "db/" prefix to the key.
So we need to strip additional 3 chars.
